### ISSUE TITLE
FeatureFlagHelper should throw its own exception

### ISSUE THREAD

**Issue by tburry** [vanilla/vanilla#7797]
This is a topic for some discussion.

Currently, `FeatureFlagHelper::throwIfNotEnabled()` takes an exception class string and arguments to throw a custom feature. This is something that we should be getting away from. Instead we should be throwing structured exceptions, possibly inherited from our general `ServerException` or `ClientException` classes.

What I recommend is something like:

```php
public static function ensureFeature(string $feature, string $message = '', $code = 403) {
    // Check for feature.
   throw new FeatureNotEnabledException($message, $code);
}
```

Something like that.

Another observation I made is that while defining feature constants on a class is nice, it does require us to load that class just to check for the feature. This may end up having us include a bunch of classes when building menus in the dashboard for example.

**Comment by initvector** [comment]
Forcing a standard exception is a good idea. The original intent was to give a lot of freedom to the classes calling the method, but in hindsight it does that to a fault. Allowing callers to dictate what exception is thrown makes it a lot harder to catch and handle in a standard way. Using a standard exception should remedy that.

I do want to keep the feature flags/configs as constants, but I'm not insistent they necessarily remain in any particular class. Addons using constants that aren't class-specific, particularly these feature flags, could be moved into a minimal (abstract) class, where the values could be referenced with bare minimum loading of additional class code (and no state). Keeping the values as constants allows for IDE hints/autocompletion, avoids errors in entering the actual flag/config string and makes it easier to update the flag/config value without having to update it across several areas of the application.

**Comment by charrondev** [comment]
Closed by #7804
